node-repository/src/main/java/com/yahoo/vespa/hosted/provision/NodesAndHosts.java

```diff
@@ -0,0 +1,58 @@
+// Copyright Yahoo. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+package com.yahoo.vespa.hosted.provision;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Wraps a NodeList and builds a host -> children mapping for faster access
+ * as that is done very frequently.
+ *
+ * @author baldersheim
+ */
+public class NodesAndHosts<NL extends NodeList> {
+    private final NL nodes;
+    private final Map<String, HostAndNodes> host2Nodes;
+
+    public static <L extends NodeList> NodesAndHosts<L> create(L nodes) {
+        return new NodesAndHosts<L>(nodes);
+    }
+
+    private NodesAndHosts(NL nodes) {
+        this.nodes = nodes;
+        host2Nodes = new HashMap<>();
+        nodes.forEach(node -> {
+            node.parentHostname().ifPresentOrElse(
+                    parent -> host2Nodes.computeIfAbsent(parent, key -> new HostAndNodes()).add(node),
+                    () -> host2Nodes.computeIfAbsent(node.hostname(), key -> new HostAndNodes()).setHost(node));
+        });
+    }
+
+    /// Return the NodeList used for construction
+    public NL nodes() { return nodes; }
+
+    public NodeList childrenOf(Node host) {
+        HostAndNodes hostAndNodes = host2Nodes.get(host.hostname());
+        return hostAndNodes != null ? NodeList.copyOf(hostAndNodes.children) : NodeList.of();
+    }
+
+    public Node parentOf(Node node) {
```

Comment:
Can we make this return `Optional<Node>`? So it's obvious to callers that it can be empty.

Reply:
Agree, done.